design-documents/graph-ql/coverage/customer-orders.md

```diff
@@ -0,0 +1,196 @@
+
+The GraphQL API should provide a possibility to retrieve orders, shipments, invoices, credit memos for the logged in customer. The current schema allows fetching only simple order details and doesn't provide a possibility to fetch order details by a number.
+
+The proposed solution is deprecation of `customerOrders` query and `orders` field to the `customer` query:
+
+```graphql
+# query to return a list of all customer orders
+type Customer {
+    orders (
+        filter: CustomerOrderFilterInput
+        currentPage: Int = 1 # current page of the customer order list. default is 1.
+        pageSize: Int = 20 # page size for the customer orders list. default is 20.
+    ): CustomerOrders
+}
+
+input CustomerOrderFilterInput {
+    order_number: String!
+}
+
+# collection of customer orders.
+type CustomerOrders {
+    items: [Order]! # collection of customer orders that contains individual order details.
+    page_info: SearchResultPageInfo
+    total_count: Int
+}
+```
+
+> Right now, we don't introduce the filter input for such entities like invoice, credit memo, shipment as all these entities are related to the same order and GraphQL resolvers anyway at first resolve the order and after that other child entities. But, in future, such filter input can be introduced as optional argument.
+
+## Order Type Schema
+
+The proposed type for the customer order might look like:
+
+```graphql
+# order details
+type Order {
+    order_date: String! # date when the order was placed.
+    status: String! # current status of the order.
+    order_number: String! # order number.
+    items: [OrderItem] # collection of all the items purchased
+    order_pricing: OrderPricingInterface! # pricing details for the order.
+    invoices: [Invoice] # invoice list for the order.
+    credit_memos: [CreditMemo] # credit memo list for the order.
+    shipments: [OrderShipment] # shipping list for the order.
+    payment_method: [PaymentMethod] # payment details for the order.
+}
+```
+
+### Order Item
+
+The order items will be presented as separate interface which will have multiple implementations for invoice, shipment and credit memo types.
+
+```graphql
+interface SalesProductInterface {
+    name: String # name of the base product.
+    sku: String! # sku of the base product.
+    url: String # url of the base product.
+    final_price: Float! # final price for the base product including all the child products and selected options.
+    discounts: [Discount] # final discount information for the base product including discounts on options and child products.
+    selected_options: [String!] # selected options for the base product. for e.g color, size etc.
+    entered_options: [KeyValue] # entered option for the base product. for e.g logo image etc.
+}
+```
+
+The `SalesProductInterface` will be implemented by the following types:
+
+```graphql
+# Order Product implementation of OrderProductInterface
+type OrderItem implements SalesProductInterface {
```

Comment:
Does not look right because usually Item has reference to a product and quantity. At the same Item is not Product.

The same question applies to other "items" implementing `SalesProductInerface`.

Reply:
What kind of reference do you like to see? `SalesProductInterface` provides `url` for a catalog product?